gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -190,18 +191,160 @@ public boolean isJobLockEnabled() {
   /**
    * Get the job state of the most recent run of the job.
    */
-  private JobState getPreviousJobState(String jobName)
-      throws IOException {
-    if (this.jobStateStore.exists(jobName, "current" + AbstractJobLauncher.JOB_STATE_STORE_TABLE_SUFFIX)) {
-      // Read the job state of the most recent run of the job
-      List<JobState> previousJobStateList =
-          this.jobStateStore.getAll(jobName, "current" + AbstractJobLauncher.JOB_STATE_STORE_TABLE_SUFFIX);
-      if (!previousJobStateList.isEmpty()) {
-        // There should be a single job state on the list if the list is not empty
-        return previousJobStateList.get(0);
+  private Map<String, JobState> getPreviousDatasetStatesByUrns(String jobName) throws IOException {
+    Path stateStorePath = new Path(this.stateStoreRootDir, jobName);
+    if (!this.stateStoreFs.exists(stateStorePath)) {
+      return ImmutableMap.of();
+    }
+
+    FileStatus[] stateStoreFileStatuses = this.stateStoreFs.listStatus(stateStorePath, new PathFilter() {
+          @Override
+          public boolean accept(Path path) {
+            return path.getName().endsWith("current" + JOB_STATE_STORE_TABLE_SUFFIX);
+          }
+        });
+
+    if (stateStoreFileStatuses == null || stateStoreFileStatuses.length == 0) {
+      return ImmutableMap.of();
+    }
+
+    Map<String, JobState> datasetStatesByUrns = Maps.newHashMap();
+    for (FileStatus stateStoreFileStatus : stateStoreFileStatuses) {
+      List<JobState> previousDatasetStates =
+          this.jobStateStore.getAll(jobName, stateStoreFileStatus.getPath().getName());
+      if (!previousDatasetStates.isEmpty()) {
+        // There should be a single dataset state on the list if the list is not empty
```

Comment:
`getAll` returns a list even though the list itself only contains a single `JobState` object for the dataset. Each dataset state (an `JobState` object) is written to a separate file.